packages/wrangler/src/inspect.ts

```diff
@@ -0,0 +1,526 @@
+          if (evt.method === "Runtime.consoleAPICalled") {
+            logConsoleMessage(
+              evt.params as Protocol.Runtime.ConsoleAPICalledEvent
+            );
+          }
+        } else {
+          // We should never get here, but who know is 2022...
+          console.error("unrecognised devtools event:", event);
+        }
+      });
+    }
+
+    ws.addEventListener("open", () => {
+      send({ method: "Runtime.enable", id: messageCounterRef.current });
+      // TODO: This doesn't actually work. Must fix.
+      send({ method: "Network.enable", id: messageCounterRef.current++ });
+
+      keepAliveInterval = setInterval(() => {
+        send({
+          method: "Runtime.getIsolateId",
+          id: messageCounterRef.current++,
+        });
+      }, 10_000);
+    });
+
+    ws.on("unexpected-response", () => {
+      console.log("waiting for connection...");
+      /**
+       * This usually means the worker is not "ready" yet
+       * so we'll just retry the connection process
+       */
+      retryRemoteWebSocketConnection();
+    });
+
+    ws.addEventListener("close", () => {
+      clearInterval(keepAliveInterval);
+    });
+
+    return () => {
+      // clean up! Let's first stop the heartbeat interval
+      clearInterval(keepAliveInterval);
+      // Then we'll send a message to the devtools instance to
+      // tell it to clear the console.
+      wsServerRef.current.clients.forEach((client) => {
+        // We could've used `localSocket` here, but
+        // then we would have had to add it to the effect
+        // change detection array, which would have made a
+        // bunch of other stuff complicated. So we'll just
+        // cycle through all of the server's connected clients
+        // (in practice, there should only be one or zero) and send
+        // the Log.clear message.
+        client.send(
+          JSON.stringify({
+            // TODO: This doesn't actually work. Must fix.
+            method: "Log.clear",
+            // we can disable the next eslint warning since
+            // we're referencing a ref that stays alive
+            // eslint-disable-next-line react-hooks/exhaustive-deps
+            id: messageCounterRef.current++,
+            params: {},
+          })
+        );
+      });
+      // Finally, we'll close the websocket
+      close();
+      // And we'll clear `remoteWebsocket`
+      setRemoteWebSocket(undefined);
+    };
+  }, [
+    props.inspectorUrl,
+    retryRemoteWebSocketConnectionSigil,
+    props.logToTerminal,
+  ]);
+
+  // This effect tracks the state changes _between_ the local
+  // and remote websockets, and handles how messages flow between them.
+  useEffect(() => {
+    /**
+     * This event listener is used for buffering messages from
+     * the remote websocket, and flushing them
+     * when the local websocket connects.
+     */
+    function remoteMessageBufferListener(event: MessageEvent) {
+      if (typeof event.data === "string") {
```

Comment:
Why not just buffer all messages, since if the socket was active we would have forwarded these messages on anyway?